File: node/mods/rustscript/lib/opcodes/checkownnftwhere.js
/**
 * Purpose: CHECKOWNNFTWHERE opcode — NFT ownership and metadata WHERE clauses.
 */

const Slip = require('./../../../../lib/saito/slip').default;

module.exports = {
  name: 'CHECKOWNNFTWHERE',
  description: `
Checks that:
1. The submitted NFT is spendable by the transaction sender (tx.from === slip2.publicKey)
2. Additional WHERE constraints hold over NFT metadata (creator, type)

Required fields:
  utxokey1, utxokey2, utxokey3
`,
  exampleScript: {
    op: 'CHECKOWNNFTWHERE',
    where: [
      {
        field: 'creator',
        operator: '==',
        value: '<publickey>'
      },
      {
        field: 'type',
        operator: '==',
        value: 'stack'
      }
    ],
    witness: {
      utxokey1: '<utxokey1>',
      utxokey2: '<utxokey2>',
      utxokey3: '<utxokey3>'
    }
  },
  schema: {
    where: 'array:clause',
    utxokey1: 'utxokey',
    utxokey2: 'utxokey',
    utxokey3: 'utxokey'
  },

  execute(node, context) {
    if (!node || typeof node !== 'object' || !context || typeof context !== 'object') {
      return false;
    }

    const witness = node.witness;
    if (!witness || typeof witness !== 'object' || Array.isArray(witness)) {
      return false;
    }

    const utxo1 = witness.utxokey1;
    const utxo2 = witness.utxokey2;
    const utxo3 = witness.utxokey3;
    if (
      typeof utxo1 !== 'string' ||
      utxo1.length === 0 ||
      typeof utxo2 !== 'string' ||
      utxo2.length === 0 ||
      typeof utxo3 !== 'string' ||
      utxo3.length === 0
    ) {
      return false;
    }

    const slip1 = Slip.fromUtxoKey(utxo1);
    const slip2 = Slip.fromUtxoKey(utxo2);
    const slip3 = Slip.fromUtxoKey(utxo3);
    if (!slip1 || !slip2 || !slip3) {
      return false;
    }

    if (!context.__opcodes) {
      context.__opcodes = {};
    }
    context.__opcodes.checkownnftwhere = {
      nft_id: utxo3.substring(0, 66).toLowerCase()
    };

    const tx = context.tx;
    if (tx && Array.isArray(tx.from) && tx.from.length > 0) {
      const sender = tx.from[0].publicKey;
      if (sender !== slip2.publicKey) {
        return false;
      }
    }

    const nft_type =
      context.app && context.app.wallet && typeof context.app.wallet.extractNFTType === 'function'
        ? context.app.wallet.extractNFTType(utxo3)
        : null;
    const creator = slip1.publicKey;

    if (Array.isArray(node.where)) {
      for (let i = 0; i < node.where.length; i += 1) {
        const clause = node.where[i];
        if (!clause || typeof clause !== 'object') {
          return false;
        }

        let lhs;
        if (clause.field === 'creator') {
          lhs = creator;
        } else if (clause.field === 'type') {
          lhs = nft_type;
        } else {
          return false;
        }

        const rhs = clause.value;
        if (clause.operator === '==') {
          if (lhs !== rhs) {
            return false;
          }
        } else if (clause.operator === '!=') {
          if (lhs === rhs) {
            return false;
          }
        } else {
          return false;
        }
      }
    }

    return true;
  }
};

File: rust/saito-core/src/core/consensus/scripting/opcodes/checkownnftwhere.rs
use crate::core::consensus::block::Block;
use crate::core::consensus::blockchain::Blockchain;
use crate::core::consensus::slip::Slip;
use crate::core::consensus::transaction::Transaction;
use crate::core::defs::{PrintForLog, SaitoUTXOSetKey};
use serde_json::{json, Value};

pub struct CheckOwnNftWhere {
    pub name: String,
    pub description: String,
    pub script: String,
    pub schema: Value,
}

fn slip_from_utxo_key_hex(hex: &str) -> Option<Slip> {
    let key = SaitoUTXOSetKey::from_hex(hex).ok()?;
    Slip::parse_slip_from_utxokey(&key).ok()
}

fn js_strict_equal(lhs: Option<&str>, rhs: &Value) -> bool {
    match lhs {
        None => rhs.is_null(),
        Some(l) => rhs.as_str() == Some(l),
    }
}

impl CheckOwnNftWhere {
    pub fn validate(
        context: &mut Value,
        tx: Option<&Transaction>,
        _blk: Option<&Block>,
        _blockchain: Option<&Blockchain>,
    ) -> u8 {
        let witness = &context["witness"];
        if !witness.is_object() {
            return 0;
        }

        let utxo1 = witness["utxokey1"].as_str().unwrap_or("");
        let utxo2 = witness["utxokey2"].as_str().unwrap_or("");
        let utxo3 = witness["utxokey3"].as_str().unwrap_or("");
        if utxo1.is_empty() || utxo2.is_empty() || utxo3.is_empty() {
            return 0;
        }

        let Some(slip1) = slip_from_utxo_key_hex(utxo1) else {
            return 0;
        };
        let Some(slip2) = slip_from_utxo_key_hex(utxo2) else {
            return 0;
        };
        let Some(_slip3) = slip_from_utxo_key_hex(utxo3) else {
            return 0;
        };

        let nft_id = utxo3
            .get(..66.min(utxo3.len()))
            .unwrap_or(utxo3)
            .to_ascii_lowercase();

        if !context
            .get("__opcodes")
            .map(|v| v.is_object())
            .unwrap_or(false)
        {
            context["__opcodes"] = json!({});
        }
        context["__opcodes"]["checkownnftwhere"] = json!({
            "nft_id": nft_id
        });

        if let Some(tx) = tx {
            if !tx.from.is_empty() {
                let sender = tx.from[0].public_key.to_base58();
                let slip2_public_key = slip2.public_key.to_base58();
                if sender != slip2_public_key {
                    return 0;
                }
            }
        }

        // context.app.wallet.extractNFTType is unavailable in the Rust evaluator.
        let nft_type: Option<&str> = None;
        let creator = slip1.public_key.to_base58();

        if let Some(where_clauses) = context["script"]["where"].as_array() {
            for clause in where_clauses {
                if !clause.is_object() {
                    return 0;
                }

                let field = clause["field"].as_str().unwrap_or("");
                let lhs = if field == "creator" {
                    Some(creator.as_str())
                } else if field == "type" {
                    nft_type
                } else {
                    return 0;
                };

                let rhs = &clause["value"];
                let operator = clause["operator"].as_str().unwrap_or("");

                match operator {
                    "==" => {
                        if !js_strict_equal(lhs, rhs) {
                            return 0;
                        }
                    }
                    "!=" => {
                        if js_strict_equal(lhs, rhs) {
                            return 0;
                        }
                    }
                    _ => return 0,
                }
            }
        }

        1
    }
}
